File: src/main/java/de/dominik48n/discordbot/DiscordBot.java
package de.dominik48n.discordbot;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import de.dominik48n.discordbot.command.CommandManager;
import de.dominik48n.discordbot.command.ShutdownCommand;
import de.dominik48n.discordbot.config.BotConfiguration;
import de.dominik48n.discordbot.config.FileConfiguration;
import de.dominik48n.discordbot.logger.Logger;
import javax.security.auth.login.LoginException;
import lombok.Getter;
import net.dv8tion.jda.api.OnlineStatus;
import net.dv8tion.jda.api.entities.Activity;
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder;
import net.dv8tion.jda.api.sharding.ShardManager;

/**
 * Created by Dominik48N on 31.01.2021
 */
@Getter
public class DiscordBot {

    @Getter private static DiscordBot instance;

    public static Gson GSON = new GsonBuilder().serializeNulls().setPrettyPrinting().disableHtmlEscaping().create();

    private final BotConfiguration botConfiguration;
    private final CommandManager commandManager;

    private ShardManager shardManager;

    public DiscordBot() {
        instance = this;

        this.botConfiguration = new FileConfiguration( this.getClass().getClassLoader() ).load();
        this.commandManager = new CommandManager();
    }

    /**
     * Start the discord bot
     */
    public void start() {
        Logger.INFO.print( "Start the bot..." );

        this.configureBot();
        this.registerCommands();
    }

    /**
     * Stops the discord bot
     */
    public void shutdown() {
        Logger.INFO.print( "Stops the bot..." );

        this.shardManager.setStatus( OnlineStatus.OFFLINE );
        this.shardManager.shutdown();

        this.commandManager.stop();

        System.exit( 0 );
    }

    /**
     * Register all commands
     */
    private void registerCommands() {
        this.commandManager.addCommands( new ShutdownCommand() );
    }

    /**
     * Configures the bot
     */
    private void configureBot() {
        final DefaultShardManagerBuilder builder = new DefaultShardManagerBuilder();

        builder.setToken( this.botConfiguration.getToken() );
        builder.setStatus( OnlineStatus.ONLINE );
        builder.setActivity( Activity.playing( this.botConfiguration.getPlayingGame() ) );

        try {
            this.shardManager = builder.build();
        } catch ( final LoginException e ) {
            e.printStackTrace();
        }
    }

}

File: src/main/kotlin/de/dominik48n/discordbot/DiscordBot.kt
package de.dominik48n.discordbot

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import de.dominik48n.discordbot.command.CommandManager
import de.dominik48n.discordbot.command.ShutdownCommand
import de.dominik48n.discordbot.config.BotConfiguration
import de.dominik48n.discordbot.config.FileConfiguration
import de.dominik48n.discordbot.logger.Logger
import net.dv8tion.jda.api.OnlineStatus
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.sharding.DefaultShardManagerBuilder
import net.dv8tion.jda.api.sharding.ShardManager
import javax.security.auth.login.LoginException
import kotlin.system.exitProcess

class DiscordBot {

    companion object {
        val GSON: Gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
    }

    private var botConfiguration: BotConfiguration = FileConfiguration(this.javaClass.classLoader).load()
    private var commandManager: CommandManager = CommandManager()

    private lateinit var shardManager: ShardManager

    /**
     * Start the discord bot
     */
    fun start() {
        Logger.INFO.print("Start the bot...")

        this.configureBot()
        this.registerCommands()
    }

    /**
     * Stops the discord bot
     */
    fun stop() {
        Logger.INFO.print("Stops the bot...")

        this.shardManager.setStatus(OnlineStatus.OFFLINE)
        this.shardManager.shutdown()

        this.commandManager.stop()

        exitProcess(0)
    }

    /**
     * Register all commands
     */
    private fun registerCommands() {
        this.commandManager.addCommand("shutdown", ShutdownCommand())
    }

    /**
     * Configures the bot
     */
    private fun configureBot() {
        val defaultSharedManagerBuilder = DefaultShardManagerBuilder()

        defaultSharedManagerBuilder.setToken(this.botConfiguration.token)
        defaultSharedManagerBuilder.setStatus(OnlineStatus.ONLINE)
        defaultSharedManagerBuilder.setActivity(Activity.playing(this.botConfiguration.playingGame))

        try {
            this.shardManager = defaultSharedManagerBuilder.build()
        } catch (e: LoginException) {
            e.printStackTrace()
        }
    }

}
